force-app/main/default/classes/DMLOperations.cls

```diff
@@ -180,28 +216,67 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
```

Comment:
This could have been a much shorter method if you followed the hint found in the method description:
`Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.`